### ISSUE TITLE
PDFs don't render correctly.

### ISSUE THREAD

**Issue by bigfatbird** [babluboy/bookworm#82]
Text is aligned oddly, code indentation isn't looking right, and i guess some characters are not encoded correctly.

**Comment by babluboy** [comment]
@bigfatbird Yes, I am aware of this issue. I'm using poppler utils at the moment to convert PDF to HTML to render the content and the conversion is not great...I will use this issue to track this and see if I can extract the text and images programatically to have greater control in rendering the content...I can also see if some css can be applied to render the text a little better...

At the current time you can use the reading preferences for line width and line height to adjust the content a little bit better...

**Comment by babluboy** [comment]
@bigfatbird Can you post a screen shot here to show how the text currently for a PDF and whether the PDF is image rich or just text...

**Comment by bigfatbird** [comment]
Sure. Here are two screenshots of the same book for example. 
http://imgur.com/a/WaZpn

**Comment by babluboy** [comment]
@bigfatbird thanks. looks like if I can center the content the rendering will look better...that should not be hard to achieve...will update here when I get to this issue

See if line width helps a little better until I get the fix in..

**Comment by bigfatbird** [comment]
Just curious: Why do you want to style it yourself, if there is an existing PDF standard? 
A pdf should look exactly like it was released, I assume.

**Comment by babluboy** [comment]
Not sure, at the moment I'm using poppler util pdftohtml and I dont see any option to render the html the way it looks like in Evince viewer...perhaps I should check the Evince code to see how the rendering is done

**Comment by bigfatbird** [comment]
Maybe you can integrate mozillas pdf.js

Von meinem iPhone gesendet

**Comment by babluboy** [comment]
Thats sounds great...thanks for the suggestion...looks workable at a quick glance..
https://mozilla.github.io/pdf.js/examples/

**Comment by unhammer** [comment]
would using on-the-fly pdf rendering instead of cached pdftohtml give a smaller bookworm db too? Mine is already up to 1.9GB

**Comment by babluboy** [comment]
@unhammer how many books are there in the library and how many are PDFs? The actual book content is cached on the file system(if the cache preference is set) but the metadata including the table of contents is stored in the db. I have seen that pdf2html generates a lot of pages as I separate the html by page break tag... Will look at better PDF handling in the future....

If you turn off caching then book content will be cached in /tmp and automatically be removed on restart... If you open the book again the same will be parsed and the html content regenerated in /tmp...this takes slightly longer to resume reading....

**Comment by unhammer** [comment]
in my case, 297 pdf's and 34 other files

**Comment by babluboy** [comment]
hmm...while 300 PDFs seem a largi-ish library (i have not tested more than 100 PDFs), yet it does feel high just for the content data to be 1.9 GB...will look into this to see if I can replicate...

**Comment by m0t0rh34d** [comment]
Render PDF content does make sense because it allows one specify how the content should look (change font, colorize something etc) which is quite problematic when it's simply PDF.

**Comment by babluboy** [comment]
@bigfatbird It dosen't look like it will be possible to extract PDF to HTML using pdf.js based on this:
[https://github.com/mozilla/pdf.js/issues/8732](https://github.com/mozilla/pdf.js/issues/8732)
Bookworm relies upon HTML files to apply all the text/color modifications, highlighting, search, navigation, etc.
I will need to either render the output of PDF2HTML in a better way or find some other way to create html pages out of PDF...

**Comment by babluboy** [comment]
Looks like poppler can be used to get the chapters from the book using this example:
https://stackoverflow.com/questions/7131906/how-to-extract-pdf-index-table-of-contents-with-poppler

At least it will reduce the data in the meta data database by just storing just the chapters and their corresponding html file. Currently i'm storing the location of all the html files which is one per page of the PDF thereby bloating the DB size as mentioned [here](https://github.com/babluboy/bookworm/issues/82#issuecomment-318816104) by @unhammer

**Comment by babluboy** [comment]
@preconf unfortunately I have not spent further time on this. I tried the following library but the extraction was too slow although the rendering was better:
https://github.com/coolwanglu/pdf2htmlEX

Will check evince to see if it is usable.

**Comment by Preconf** [comment]
Sounds good, yes Evince seems to be rendering PDFs correctly (is what I end up using until bookworm works) so using it as a reference is a good idea.

**Comment by prog-amateur** [comment]
> @Preconf unfortunately I have not spent further time on this. I tried the following library but the extraction was too slow although the rendering was better:
> https://github.com/coolwanglu/pdf2htmlEX
> 
> Will check evince to see if it is usable.

Hello, I came here after a review of your app in a website. Everything is perfect, except this PDF support : it should be as the original one, not re-arranged. This is clearly a No-Go for this specific aspect, and people can have many PDF in their library. 

So do you have any solution please ? this issue was open in 2017. 

Thank you, and please remember : I claim here but I really like this app.

**Comment by CrisBRM** [comment]
Completely unusable for pdfs. Might as well just stick to koreader for the epub support and evince for everything else.

**Comment by prog-amateur** [comment]
> Completely unusable for pdfs. Might as well just stick to koreader for the epub support and evince for everything else.

Be careful, you can have a thumb down (like for my request) for saying that this app cannot read PDF correctly. There is some people here who maybe want to stay with a pdf code reader ( @bigfatbird  ?)

**Comment by bigfatbird** [comment]
@prog-amateur I downvoted your and the other reply as they are completely useless for fixing the bug and just create unnecessary work. A "it doesn't work for me, too" under a bug which just describes that something is not working as expected is not helpful at all and is more work for the developers. It was not personal.
